File: packages/backend/src/middleware/jwtAuth.ts
/**
 * JWT 认证中间件
 */

import { Request, Response, NextFunction } from 'express'
import jwt, { SignOptions } from 'jsonwebtoken'
import { createLogger } from '../utils/logger.js'
import { getConfig } from '../config/index.js'
import { isTokenValid } from '../storage/tokenStore.js'

const logger = createLogger('JwtAuthMiddleware')

// JWT Payload 接口
export interface JwtPayload {
  username: string
  iat: number
  exp: number
}

// 扩展 Express Request 类型
declare global {
  namespace Express {
    interface Request {
      user?: JwtPayload
    }
  }
}

// 不需要认证的路径（相对于挂载点，如 /api 下的 /auth/login）
const PUBLIC_PATHS = [
  '/auth/login',
  '/health',
  '/admin/health'
]

/**
 * 检查路径是否为公开路径（不需要认证）
 */
function isPublicPath(path: string): boolean {
  return PUBLIC_PATHS.some(publicPath => path === publicPath || path.startsWith(publicPath + '/'))
}

/**
 * JWT 认证中间件
 * 验证 JWT Token，排除公开路径
 */
export async function jwtAuthMiddleware(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  // 公开路径不需要认证
  if (isPublicPath(req.path)) {
    next()
    return
  }

  const config = getConfig()

  // 提取 Token
  let token: string | undefined

  // 从 Authorization header 提取 (Bearer token)
  const authHeader = req.headers.authorization
  if (authHeader?.startsWith('Bearer ')) {
    token = authHeader.substring(7)
  }

  if (!token) {
    logger.warn('Missing JWT token', { path: req.path, ip: req.ip })
    res.status(401).json({
      error: {
        message: 'Authentication required',
        type: 'authentication_error'
      }
    })
    return
  }

  try {
    // 验证 Token
    const decoded = jwt.verify(token, config.jwt.secret) as JwtPayload

    // 检查 Token 是否存在于 Redis（白名单验证）
    const valid = await isTokenValid(token)
    if (!valid) {
      logger.warn('Token not found in Redis (logged out or expired)', { path: req.path, ip: req.ip })
      res.status(401).json({
        error: {
          message: 'Token has been revoked',
          type: 'authentication_error'
        }
      })
      return
    }

    // 将用户信息附加到请求对象
    req.user = decoded

    next()
  } catch (error) {
    if (error instanceof jwt.TokenExpiredError) {
      logger.warn('JWT token expired', { path: req.path, ip: req.ip })
      res.status(401).json({
        error: {
          message: 'Token expired',
          type: 'authentication_error'
        }
      })
      return
    }

    if (error instanceof jwt.JsonWebTokenError) {
      logger.warn('Invalid JWT token', { path: req.path, ip: req.ip, error: (error as Error).message })
      res.status(401).json({
        error: {
          message: 'Invalid token',
          type: 'authentication_error'
        }
      })
      return
    }

    logger.error('JWT verification error', { path: req.path, error: (error as Error).message })
    res.status(500).json({
      error: {
        message: 'Authentication error',
        type: 'server_error'
      }
    })
  }
}

/**
 * 生成 JWT Token
 */
export function generateToken(username: string): string {
  const config = getConfig()

  // 使用类型断言解决 expiresIn 类型兼容性问题
  return jwt.sign(
    { username },
    config.jwt.secret,
    { expiresIn: config.jwt.expiresIn } as SignOptions
  )
}

/**
 * 解析 JWT Token（不验证）
 */
export function decodeToken(token: string): JwtPayload | null {
  try {
    return jwt.decode(token) as JwtPayload
  } catch {
    return null
  }
}

File: packages/gateway/middleware/jwt_auth.go
package middleware

import (
	"context"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/golang-jwt/jwt/v5"
	"github.com/rs/zerolog/log"

	"github.com/kiro-gateway/gateway/config"
	"github.com/kiro-gateway/gateway/storage"
	"github.com/kiro-gateway/gateway/util"
)

// publicPaths lists paths that do not require JWT authentication.
var publicPaths = []string{
	"/api/auth/login",
	"/health",
	"/api/admin/health",
}

// JWTAuthMiddleware returns a Gin middleware that validates JWT tokens from
// the Authorization header. Tokens must also exist in the Redis whitelist.
func JWTAuthMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		path := c.Request.URL.Path

		// Skip authentication for public paths.
		for _, pp := range publicPaths {
			if path == pp {
				c.Next()
				return
			}
		}

		// Extract token from Authorization header.
		authHeader := c.GetHeader("Authorization")
		if authHeader == "" {
			util.UnauthorizedError(c, "Authorization header is required")
			c.Abort()
			return
		}

		parts := strings.SplitN(authHeader, " ", 2)
		if len(parts) != 2 || !strings.EqualFold(parts[0], "Bearer") {
			util.UnauthorizedError(c, "Invalid authorization format, expected Bearer token")
			c.Abort()
			return
		}
		tokenString := parts[1]

		// Parse and validate the JWT.
		cfg := config.GetConfig()
		token, err := jwt.Parse(tokenString, func(t *jwt.Token) (interface{}, error) {
			if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
				return nil, jwt.ErrSignatureInvalid
			}
			return []byte(cfg.JWT.Secret), nil
		})
		if err != nil || !token.Valid {
			util.UnauthorizedError(c, "Invalid or expired token")
			c.Abort()
			return
		}

		// Verify token exists in Redis whitelist.
		ctx := context.Background()
		valid, err := storage.IsTokenValid(ctx, tokenString)
		if err != nil {
			log.Error().Err(err).Msg("Failed to validate token in Redis")
			util.ErrorJSON(c, http.StatusInternalServerError,
				"Token validation failed", util.ErrTypeServer)
			c.Abort()
			return
		}
		if !valid {
			util.UnauthorizedError(c, "Token has been revoked")
			c.Abort()
			return
		}

		// Extract username from claims and set on context.
		subject, err := token.Claims.GetSubject()
		if err != nil || subject == "" {
			util.UnauthorizedError(c, "Invalid token claims")
			c.Abort()
			return
		}

		c.Set("username", subject)
		c.Set("token", tokenString)
		c.Next()
	}
}

// GenerateToken creates a signed JWT for the given username and stores it in
// the Redis whitelist with the configured TTL.
func GenerateToken(username string) (string, error) {
	cfg := config.GetConfig()
	expiresIn := ParseJWTExpiresIn(cfg.JWT.ExpiresIn)

	claims := jwt.RegisteredClaims{
		Subject:   username,
		ExpiresAt: jwt.NewNumericDate(time.Now().Add(expiresIn)),
		IssuedAt:  jwt.NewNumericDate(time.Now()),
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	tokenString, err := token.SignedString([]byte(cfg.JWT.Secret))
	if err != nil {
		return "", err
	}

	// Store token in Redis whitelist.
	ttlSeconds := int(expiresIn.Seconds())
	ctx := context.Background()
	if err := storage.StoreToken(ctx, tokenString, username, ttlSeconds); err != nil {
		log.Error().Err(err).Msg("Failed to store token in Redis")
		return "", err
	}

	return tokenString, nil
}

// ParseJWTExpiresIn converts a duration string like "24h", "7d", or "1h" to a
// time.Duration. Falls back to 24 hours if the string cannot be parsed.
func ParseJWTExpiresIn(expiresIn string) time.Duration {
	expiresIn = strings.TrimSpace(expiresIn)
	if expiresIn == "" {
		return 24 * time.Hour
	}

	// Handle "d" suffix for days.
	if strings.HasSuffix(expiresIn, "d") {
		numStr := strings.TrimSuffix(expiresIn, "d")
		days, err := strconv.Atoi(numStr)
		if err == nil && days > 0 {
			return time.Duration(days) * 24 * time.Hour
		}
		return 24 * time.Hour
	}

	// Delegate to Go's standard duration parser for "h", "m", "s", etc.
	d, err := time.ParseDuration(expiresIn)
	if err != nil {
		return 24 * time.Hour
	}
	return d
}
